Adds two more Bluetooth samples

yeshvanth-m opened the pull request:
The scan_adv sample has combined broadcaster and observer, I've created two separate samples, one to demonstrate the role of Broadcaster (Advertiser) and the other to demonstrate the role of Observer (Scanner) which would be helpful for beginners.

## Screenshots
### Terminal output of Advertiser Sample
![Advertiser_Terminal](https://user-images.githubusercontent.com/73386649/140162188-a7d75e18-dbaf-4663-a1ff-56706183fab5.PNG)

### nRF Connect Mobile App for advertiser sample
![nRF_Connect_App](https://user-images.githubusercontent.com/73386649/140162195-e541497b-5f84-4e47-aa38-9761e40cdf55.png)

### Terminal output of Scanner Sample
![Scanner_Terminal](https://user-images.githubusercontent.com/73386649/140162198-72869ce1-61c4-4e3b-8b1a-9e339338dcce.PNG)


Signed-off-by: Yeshvanth M <yeshvanthmuniraj@gmail.com>

yeshvanth-m commented:
@jhedberg and @Vudentz, please let me know if I need to make any changes.

yeshvanth-m commented:
Hi @alwa-nordic @sjanc  @thoh-ot @ppryga and @cvinayak would like to know your feedback :)
Thanks 
Yeshvanth

cvinayak commented on samples/bluetooth/scanner/src/main.c:
```diff
@@ -0,0 +1,69 @@
+	struct bt_le_scan_param scan_param = {
+		.type       = BT_HCI_LE_SCAN_PASSIVE,
+		.options    = BT_LE_SCAN_OPT_NONE,
+		.interval   = 0x0010,
+		.window     = 0x0010,
```
Use define from gap.h file.

yeshvanth-m replied:
Yeah, I've used the defines from gap.h file

cvinayak commented on samples/bluetooth/scanner/src/main.c:
```diff
@@ -0,0 +1,69 @@
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+        int i, flag = 0;
+        for(i = 0; i < pck_len; i++) {
+                if(!(strcmp(devices[i], addr_str))) {
+                        flag = 1; // Already present
+                }
+        }
+        if(!flag) {
+                strcpy(devices[i], addr_str);
+                pck_len++;
+                printk("Device found: %s (RSSI %d) \n", addr_str, rssi);
+        }
```
You should use the duplicate filtering feature in the controller, by supplying the `options` in the `scan_param`

yeshvanth-m replied:
Yes, I have used the filter option.

cvinayak commented on samples/bluetooth/advertiser/src/main.c:
```diff
@@ -0,0 +1,62 @@
+	do {
+		k_msleep(1000);
+
+                 printk("Sending advertising data: 0x%02X \n", mfg_data[2]);
```
Use tabs for indentations everywhere.

yeshvanth-m replied:
Corrected indentation.

cvinayak reviewed (changes requested):
For advertiser and scanner states, use the GAP names `broadcaster` and `observer`, respectively.

tejlmand commented on samples/bluetooth/advertiser/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
```
Nothing wrong on this line, but the commit message is just this:
>  Adds two more Bluetooth samples
>
> Signed-off-by: Yeshvanth M <yeshvanthmuniraj@gmail.com>

That is simply insufficient for a commit message, especially if looking at this during a bisect.

Please write some more info, for example:
- What sample is added
- What can this sample do
- Anything else that is special or valuable to know about this sample

Please note, the commit message is also an aid to the reviewers in order to understand the code and why it is being added.

yeshvanth-m replied:
I've added the required details to the commit message.

tejlmand commented on samples/bluetooth/advertiser/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
CMake 3.20 is now the minimum version required by Zephyr.
```suggestion
cmake_minimum_required(VERSION 3.20)
```

yeshvanth-m replied:
Changed the cmake minimum version requirement.

tejlmand commented on samples/bluetooth/scanner/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
CMake 3.20 is now the minimum version required by Zephyr.

```suggestion
cmake_minimum_required(VERSION 3.20)
```

tejlmand commented on samples/bluetooth/advertiser/boards/rv32m1_vega_ri5cy.overlay:
```diff
@@ -0,0 +1,22 @@
+/*
```
just being curious, no changes requested.

I see several bluetooth samples, but not all, having this dts overlay file.

Do you know when this file is required ?
Or are you just including this file because it was present in the sample you used as template.

yeshvanth-m replied:
Yeah, I use nRF5340 DK as my target. Even I'm not sure when this overlay would be used.

tejlmand commented on samples/bluetooth/advertiser/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+sample:
+  description: TBD
+  name: TBD
```
So when do you intend to write something here, if not now ?

tejlmand commented on samples/bluetooth/scanner/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+sample:
+  description: TBD
+  name: TBD
```
So when do you intend to write something here, if not now ?

yeshvanth-m replied:
I have given the name, removed the description field as it's already available in the README file.

tejlmand reviewed (changes requested):
Some small comments.

Also it would be nice if each sample is added as an independent commit, imho, but not hard requirement.
They can still be part of same PR.

tejlmand commented:
@yeshvanth-m please take a look at your commits.

Also take a look here: https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow
Bullet 15 contains a small suggestion on how to work on review comments.

yeshvanth-m commented:
Hi @tejlmand,

I forgot to add the signature at the end of the commit message. So I edited the commit message and force pushed to the add/bluetooth_samples branch. However, I feel starting a fresh PR with all changes as requested by yourself and @cvinayak would be easier.  
**Shall I open a new PR?** 

I've changed the codes as requested in the latest commit. Please let me know if the code is okay.
I'm attaching the commit messages for both the samples that I'll be adding in this comment, please review it and let me know if it is fine.

**Commit message for Observer Sample:**

samples: Bluetooth: Added Observer sample 

Added Observer sample

This sample scans for nearby BLE devices and prints the address of the device found and the RSSI value to the UART terminal.

Uses passive scanning and demonstrates the role of the observer.

Signed-off-by: Yeshvanth M <yeshvanthmuniraj@gmail.com>


**Commit message for Broadcaster Sample:**

samples: Bluetooth: Added Broadcaster sample 

This sample periodically sends out advertising data packets with the manufacturer data element.

The content of data sent is a single byte indicating how many advertising packets the device has sent which rolls back to 0 after 255.
Demonstrates role of broadcaster.

Signed-off-by: Yeshvanth M <yeshvanthmuniraj@gmail.com>

tejlmand commented:
> Shall I open a new PR?

That's not necessary.
I think it would be better continuing this PR, that way history / comments are preserved.

> I'm attaching the commit messages for both the samples that I'll be adding in this comment, please review it and let me know if it is fine.

If you force push you branch then we'll be able to see the updated commit messages there.

yeshvanth-m commented:
Hi @tejlmand,

I've modified the commit messages and force-pushed them as you said.

Please review it and let me know.

Thanks

tejlmand commented:
> I've modified the commit messages and force-pushed them as you said.
> 
> Please review it and let me know.
> 
> Thanks

Thanks for cleaning it up.
You're almost there, but there are still three commits, so try to squash the content of this commit https://github.com/zephyrproject-rtos/zephyr/pull/40057/commits/3c8bcf8679a7f408d97f297ef6b4a3c647175763 into the appropriate new individual commits.

yeshvanth-m commented:
Hi @tejlmand,

Let me know if I've understood the point correctly:

I should squash the commit 3c8bcf8 into commit 41c3627 and leave the commit  e208123 as it is.

If not, please clarify on this comment.
>You're almost there, but there are still three commits, so try to squash the content of this commit 3c8bcf8 into the appropriate >new individual commits.

tejlmand commented:
> I should squash the commit [3c8bcf8](https://github.com/zephyrproject-rtos/zephyr/commit/3c8bcf8679a7f408d97f297ef6b4a3c647175763) into commit [41c3627](https://github.com/zephyrproject-rtos/zephyr/commit/41c3627c7b83ca8c96b24a03bd6bbf62deebf62f) and leave the commit [e208123](https://github.com/zephyrproject-rtos/zephyr/commit/e208123c14edbe44e9a0d04b965081735558c4e2) as it is.
> 
> If not, please clarify on this comment.

What I meant is the first commit should be split into the two samples.

This can be achieved this way:
```
$ git checkout add/bluetooth_samples
$ git rebase -i HEAD~3
# Mark all three commits for `edit`

$ git reset HEAD^
# This resets head and prepare for next step

$ git add samples/bluetooth/advertiser/ 
$ git commit -m advertiser

$ git add samples/bluetooth/scanner/
$ git commit -m scanner

$ git rebase --continue
$ git commit --amend --squash HEAD~1
$ git rebase --continue 
$ git commit --amend --squash HEAD~3
$ git rebase --continue
# Now first rebase is done, and there are now four commits that we can autosquash

$ git rebase -i --autosquash HEAD~4
# Git already re-ordered and marked the correct commits for squashing, so just save and close.

# Git will now open your editor for commit message adjustment, just use the good commit messages you already have.
# Remember to adjust the line length (72 char limit), see ref below.
# Save and close editor

# Push your cleaned up commit.
```

That should give you this:
https://github.com/tejlmand/zephyr/commits/review/40057_commit_cleanup

Two nice commits with clean history.

*ref*: https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines

yeshvanth-m committed:
samples: Bluetooth: Added Broadcaster sample

This sample periodically sends out advertising data packets with
manufacturer data element.

The content of data sent is a single byte indicating how many
advertising packets the device has sent which rolls back to 0 after 255.
Demonstrates role of broadcaster.

Signed-off-by: Yeshvanth M <yeshvanthmuniraj@gmail.com>

yeshvanth-m committed:
samples: Bluetooth: Added Observer sample

This sample scans for nearby BLE devices and prints the address of the
device found and the RSSI value to the UART terminal.

Uses passive scanning and demonstrates the role of the observer.

Signed-off-by: Yeshvanth M <yeshvanthmuniraj@gmail.com>

yeshvanth-m commented:
Hi @tejlmand 

I've pushed the cleaned-up commits and pushed it as you mentioned.  Please review it and let me know if any more changes are required.

Thank you for the guidance in the auto squashing process :)

tejlmand commented:
> Thank you for the guidance in the auto squashing process :)

Welcome.

tejlmand reviewed (approved):
lgtm

yeshvanth-m commented:
Hi @cvinayak

I have done the requested changes. Please review it and let me know if any more changes are required :)

yeshvanth-m commented:
Hi @cvinayak 

I have done the requested changes. Please review it and let me know if any more changes are required.

cvinayak commented on samples/bluetooth/broadcaster/README.rst:
```diff
@@ -0,0 +1,27 @@
+Overview
+********
+
+A simple application demonstrating BLE Broadcaster role functionality. 
```
```suggestion
A simple application demonstrating Bluetooth Low Energy Broadcaster role functionality. 
```

cvinayak commented on samples/bluetooth/broadcaster/src/main.c:
```diff
@@ -0,0 +1,62 @@
+{
+	int err;
+
+	printk("Starting Advertiser\n");
```
```suggestion
	printk("Starting Broadcaster\n");
```

cvinayak commented on samples/bluetooth/broadcaster/src/main.c:
```diff
@@ -0,0 +1,62 @@
+	printk("Bluetooth initialized\n");
+
+	do {
+		k_msleep(1000);
+
+                printk("Sending advertising data: 0x%02X \n", mfg_data[2]);
```
Indentation, use tabs,

```suggestion
		k_msleep(1000);

		printk("Sending advertising data: 0x%02X \n", mfg_data[2]);
```

cvinayak commented on samples/bluetooth/broadcaster/src/main.c:
```diff
@@ -0,0 +1,62 @@
+			return;
+		}
+            
+                mfg_data[2]++;
```
Indentation, use tabs.
```suggestion
		mfg_data[2]++;
```

cvinayak commented on samples/bluetooth/observer/README.rst:
```diff
@@ -0,0 +1,25 @@
+Overview
+********
+
+A simple application demonstrating BLE Observer role functionality. 
```
```suggestion
A simple application demonstrating Bluetooth Low Energy Observer role functionality. 
```

cvinayak commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -0,0 +1,54 @@
+		printk("Starting scanning failed (err %d)\n", err);
+		return;
+	}
+        
```
Redundant newline and indentation.
```suggestion
```

yeshvanth-m commented:
Hi @cvinayak 

I've done the requested changes.

jhedberg reviewed (changes requested):
@yeshvanth-m could you please squash the fixes in "samples: Bluetooth: Fix indentation and names" to the appropriate two other commits? If a commit hasn't been merged yet and has an issue in it the issue should be fixed in the commit itself rather than follow up commits in the same (unmerged) PR.

You can do this e.g. as follows:
```
git reset HEAD~1                # Eliminate the last commit, but retain the changes in the tree
git add -i                      # Add the chunks that should belong to one of the commits (e.g. Observer)
git commit -m "fix observer"    # Create a temporary commit with the observer fixes
git add -i                      # Add the chunks that should belong to the other commit (e.g. Broadcaster)
git commit -m "fix broadcaster" # Create a temporary commit with the broadcaster fixes
git rebase -i                   # Edit the commit list and move the "fixup" commits right under the commit
                                # that they are fixing, and change the first letter of the line to "f"
```

yeshvanth-m commented:
Hi @jhedberg 

I've squashed the fixes in "samples: Bluetooth: Fix indentation and names" into the appropriate commits. Please review it and let me know. 

Thanks for the guidance in fixup :)

Hi @cvinayak 

I've done the changes that you requested too. Please review it.

cvinayak commented on samples/bluetooth/observer/src/main.c:
```diff
@@ -0,0 +1,53 @@
+	};
+	int err;
+
+	printk("Starting Scanner\n");
```
One minor change...
```suggestion
	printk("Starting Observer\n");
```

cvinayak reviewed (changes requested):
One last change requested, otherwise good work.

cvinayak reviewed (approved):
LGTM (not tested myself on target though)

yeshvanth-m commented:
Hi @carlescufi 

Please review the PR and let me know if any changes are needed :)

jhedberg commented:
@yeshvanth-m this cannot be merged until you have fixed the failing Compliance issues.

yeshvanth-m commented:
Hi @jhedberg 

I made the changes and pushed it, but getting an error while running tests with the twister.

![image](https://user-images.githubusercontent.com/73386649/143376839-a43b51c7-f8c5-4a8b-8708-ba929b97f9e6.png)

Should I run the following commands and push them again?

```
git fetch -all
git rebase --ignore-whitespace upstream/main
git rebase -i <commit-id>
git add [files]
git rebase --continue
git push --force origin add/bluetooth_samples
```

If not, please let me know what I need to do. 

Thanks in advance.

tejlmand commented:
> ```
> git fetch --all
> git rebase --ignore-whitespace upstream/main
> git push --force origin add/bluetooth_samples
> ```

This should be sufficient to do.

yeshvanth-m commented:
Hi @tejlmand 

Thanks for the info, also I'm not able to figure out why there is still some error in the compliance check.

![image](https://user-images.githubusercontent.com/73386649/143420671-015d831b-a0f7-41ea-baba-c2fe1db6faaf.png)

Please let me know how to fix this.

jhedberg commented:
> Will this be fixed if I remove the space before the '\n' inside the print statement argument?

No. You have unnecessary space characters after `;`

yeshvanth-m commented:
Thank you @tejlmand, @cvinayak, and @jhedberg for your guidance throughout the PR :)
